{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import itertools\n",
    "\n",
    "import pandas as pd\n",
    "from collections import Counter, defaultdict, OrderedDict\n",
    "from string import ascii_lowercase\n",
    "\n",
    "from Bio import AlignIO, SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import MutableSeq, translate\n",
    "\n",
    "from ind import findEnds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_dna_subs(read, ref, verbose=False, start_offset=0, end_trail=0):\n",
    "    \"\"\"\n",
    "    @ read, ref: MutableSeq objects\n",
    "    :return errors - tuple (position, expected triplet, actual triplet, ) / none if broken read\n",
    "    \n",
    "    This is intended purely for counting PCR polymerase substitutions.\n",
    "    The assumption is that the reference includes 0 nt either side of the gene of interest.\n",
    "    As for HGVS, the starting offset and number of trailing nt are variable\n",
    "    Letter by letter report mutations in NGS read, all counts 1- based in result (code in 0-count).\n",
    "    - substitution: 78C = nt 78 in reference is changed to C\n",
    "    - deletions: 78d6 = 6 nt deleted starting with 78: 1-77, d6, 84-end\n",
    "    - insertion: 78iATC = after nt 78 inserted seq. ATC\n",
    "    \"\"\"\n",
    "\n",
    "    if read is None:\n",
    "        if verbose:\n",
    "            print('no read provided')\n",
    "        return\n",
    "\n",
    "#     # find the ends of the read\n",
    "    ends = findEnds(read, ref, start_offset)\n",
    "\n",
    "    # scan read & reference letter by letter, counting position in reference\n",
    "    # reads have been trimmed so that reference starts @ offset=3 by default (0,1,2 is the extra triplet)\n",
    "    dna_errors = []\n",
    "    ref_index = ends.get('start') - start_offset\n",
    "    i = ends.get('start')\n",
    "    max_i = len(ref) - end_trail\n",
    "    \n",
    "    # check for indels\n",
    "    if str(read).find('-') != -1:\n",
    "        if verbose:\n",
    "            print('indel')\n",
    "        return\n",
    "    elif str(ref).find('-') != -1:\n",
    "        if verbose:\n",
    "            print('indel')\n",
    "        return\n",
    "    elif len(ref) != len(read):\n",
    "        if verbose:\n",
    "            print('indel')\n",
    "        return\n",
    "    else:\n",
    "        # continue on to scand the read for substitutions\n",
    "        while i < ends.get('end'):\n",
    "            if i > max_i:\n",
    "                break\n",
    "            # check for differences\n",
    "            if read[i] == ref[i]:\n",
    "                ref_index += 1\n",
    "                i += 1\n",
    "\n",
    "            else:\n",
    "                # substitution\n",
    "                if ref_index > 0:\n",
    "                    dna_errors += [(str(ref_index + 1), 's', str(read[i]))]\n",
    "                i += 1\n",
    "                ref_index += 1\n",
    "\n",
    "\n",
    "    return tuple(dna_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_ref = \"ATATG\"\n",
    "test_seq = \"ATTTG\"\n",
    "\n",
    "d = find_dna_subs(test_seq, test_ref)\n",
    "print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"../input_fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(wt_fw)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(wt_fw)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The fixed length of sequences means that pre-aligning the sequences is not necessary. I can directly compare each FASTA read with the reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# let's analyse to all forward sequences\n",
    "with open(\"./TEV_fw.fa\", \"r\") as f:\n",
    "    ref_seq =  SeqIO.read(f, \"fasta\", alphabet=IUPAC.ambiguous_dna)\n",
    "\n",
    "files = glob.glob(\"/home/mp/InDelScanner/input_fasta/*_fw*\")\n",
    "\n",
    "dna_variant_counts = {}\n",
    "\n",
    "for f in files:\n",
    "    # make a dictionary to collect the counts\n",
    "    fname = f[34:]\n",
    "    dna_variant_counts[fname] = defaultdict(int)\n",
    "    \n",
    "    for var_record in SeqIO.parse(f, \"fasta\", alphabet=IUPAC.ambiguous_dna):\n",
    "        dna_d = find_dna_subs(var_record.seq, ref_seq.seq, start_offset = 0, end_trail = 0)\n",
    "        if dna_d is None:\n",
    "            print(var_record)\n",
    "        else:\n",
    "            dna_variant_counts[fname][len(dna_d)] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_mutations  = max([max(dna_variant_counts[i].keys()) for i in dna_variant_counts.keys()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_order = ['H_fw_1.fa', 'H_fw_3.fa', 'H_fw_5.fa', \n",
    "                'H_fw_7.fa', 'H_fw_9.fa', 'H_fw_11.fa', \n",
    "                'L_fw_1.fa','L_fw_3.fa', 'L_fw_5.fa', \n",
    "                'L_fw_7.fa', 'L_fw_9.fa', 'L_fw_11.fa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_fw_1.fa</th>\n",
       "      <th>H_fw_3.fa</th>\n",
       "      <th>H_fw_5.fa</th>\n",
       "      <th>H_fw_7.fa</th>\n",
       "      <th>H_fw_9.fa</th>\n",
       "      <th>H_fw_11.fa</th>\n",
       "      <th>L_fw_1.fa</th>\n",
       "      <th>L_fw_3.fa</th>\n",
       "      <th>L_fw_5.fa</th>\n",
       "      <th>L_fw_7.fa</th>\n",
       "      <th>L_fw_9.fa</th>\n",
       "      <th>L_fw_11.fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>19</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    H_fw_1.fa  H_fw_3.fa  H_fw_5.fa  H_fw_7.fa  H_fw_9.fa  H_fw_11.fa  \\\n",
       "0          18          0          0          0          0           0   \n",
       "1          23          4          0          0          1           0   \n",
       "2          24          9          0          0          3           0   \n",
       "3          13         20          4          2          2           0   \n",
       "4           3         23          3          0          3           0   \n",
       "5           5         15         11          7          4           2   \n",
       "6           2          8         10          8         12           3   \n",
       "7           0          7         13         11         12           7   \n",
       "8           0          2         17         16         13          10   \n",
       "9           0          3         12         17         10          14   \n",
       "10          0          1          3         14          6          16   \n",
       "11          0          1          2          4          9          14   \n",
       "12          0          0          2          7          8           8   \n",
       "13          0          0          0          3          4           8   \n",
       "14          0          0          0          0          1           4   \n",
       "15          0          0          0          0          3           6   \n",
       "16          0          0          0          1          1           2   \n",
       "17          0          0          0          1          0           0   \n",
       "18          0          0          0          0          0           2   \n",
       "\n",
       "    L_fw_1.fa  L_fw_3.fa  L_fw_5.fa  L_fw_7.fa  L_fw_9.fa  L_fw_11.fa  \n",
       "0          41         17         10          0          0           2  \n",
       "1          36         25         13          0          0           1  \n",
       "2           7         20         16          1          0           0  \n",
       "3           4         14         19          3          8           4  \n",
       "4           1         10         10         16         23           6  \n",
       "5           0          0         14         27         19           9  \n",
       "6           1          1          5         15         13          25  \n",
       "7           0          1          1         16         12          14  \n",
       "8           0          1          1         10          4           9  \n",
       "9           0          0          0          5          5          13  \n",
       "10          0          0          1          0          6           2  \n",
       "11          0          0          0          0          3           2  \n",
       "12          0          0          0          0          1           5  \n",
       "13          0          0          0          0          0           2  \n",
       "14          0          0          0          0          0           1  \n",
       "15          0          0          0          0          0           2  \n",
       "16          0          0          0          0          0           0  \n",
       "17          0          0          0          0          0           0  \n",
       "18          0          0          0          0          0           0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = OrderedDict()\n",
    "\n",
    "for fname in column_order:\n",
    "    df[fname] = [dna_variant_counts[fname][i] for i in range(max_mutations +1)]\n",
    "    \n",
    "pd.DataFrame.from_dict(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's analyse to all second half sequences\n",
    "with open(\"./TEV_rv.fa\", \"r\") as f:\n",
    "    wt_end =  SeqIO.read(f, \"fasta\", alphabet=IUPAC.ambiguous_dna)\n",
    "\n",
    "dna_variant_counts = {}\n",
    "\n",
    "files = glob.glob(\"/home/mp/InDelScanner/input_fasta/*_rv*\")\n",
    "\n",
    "for f in files:\n",
    "    # make a dictionary to collect the counts\n",
    "    fname = f[34:]\n",
    "    dna_variant_counts[fname] = defaultdict(int)\n",
    "    \n",
    "    for var_record in SeqIO.parse(f, \"fasta\", alphabet=IUPAC.ambiguous_dna):\n",
    "        dna_d = find_dna_subs(var_record.seq, wt_end.seq, start_offset = 0, end_trail = 0)\n",
    "        if dna_d is None:\n",
    "            print(var.record)\n",
    "        else:\n",
    "            dna_variant_counts[fname][len(dna_d)] += 1\n",
    "\n",
    "observed_counts = [list(dna_variant_counts[i].keys()) for i in dna_variant_counts.keys()]\n",
    "max_mutations = max(list(itertools.chain.from_iterable(observed_counts))) #  this tolerates an empty input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['H_rv_1.fa', 'H_rv_11.fa', 'H_rv_3.fa', 'H_rv_5.fa', 'H_rv_7.fa', 'H_rv_9.fa', 'L_rv_1.fa', 'L_rv_11.fa', 'L_rv_3.fa', 'L_rv_5.fa', 'L_rv_7.fa', 'L_rv_9.fa', 'TEV_rv.fa'])"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_variant_counts.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_order = ['H_rv_1.fa', 'H_rv_3.fa','H_rv_5.fa',\n",
    "                'H_rv_7.fa', 'H_rv_9.fa','H_rv_11.fa',\n",
    "                'L_rv_1.fa', 'L_rv_3.fa', 'L_rv_5.fa',\n",
    "                'L_rv_7.fa', 'L_rv_9.fa', 'L_rv_11.fa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H_rv_1.fa</th>\n",
       "      <th>H_rv_3.fa</th>\n",
       "      <th>H_rv_5.fa</th>\n",
       "      <th>H_rv_7.fa</th>\n",
       "      <th>H_rv_9.fa</th>\n",
       "      <th>H_rv_11.fa</th>\n",
       "      <th>L_rv_1.fa</th>\n",
       "      <th>L_rv_3.fa</th>\n",
       "      <th>L_rv_5.fa</th>\n",
       "      <th>L_rv_7.fa</th>\n",
       "      <th>L_rv_9.fa</th>\n",
       "      <th>L_rv_11.fa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>28</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>16</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    H_rv_1.fa  H_rv_3.fa  H_rv_5.fa  H_rv_7.fa  H_rv_9.fa  H_rv_11.fa  \\\n",
       "0          14          0          0          0          0           0   \n",
       "1          25          4          0          0          0           0   \n",
       "2          27          4          0          0          0           0   \n",
       "3          11          7          0          0          0           0   \n",
       "4           8         12          2          0          3           0   \n",
       "5           7         18          2          0          1           0   \n",
       "6           0         12          6          0          2           3   \n",
       "7           0         13          9          0          5           2   \n",
       "8           0         11         15          0          5           8   \n",
       "9           0          8         20          0          8           5   \n",
       "10          0          3         15          0         11          12   \n",
       "11          0          2          8          0          8           7   \n",
       "12          0          1          3          0         12           7   \n",
       "13          0          0          7          0          9           7   \n",
       "14          0          0          4          0          9          17   \n",
       "15          0          0          3          0          6           6   \n",
       "16          0          0          0          0          5           9   \n",
       "17          0          0          0          0          4           5   \n",
       "18          0          0          0          0          5           3   \n",
       "19          0          0          0          0          1           1   \n",
       "20          0          0          0          0          1           1   \n",
       "21          0          0          0          0          0           1   \n",
       "22          0          0          0          0          0           0   \n",
       "23          0          0          0          0          0           0   \n",
       "24          0          0          0          0          0           0   \n",
       "25          0          0          0          0          0           0   \n",
       "26          0          0          0          0          0           0   \n",
       "27          0          0          0          0          0           1   \n",
       "\n",
       "    L_rv_1.fa  L_rv_3.fa  L_rv_5.fa  L_rv_7.fa  L_rv_9.fa  L_rv_11.fa  \n",
       "0          48         13         11          0          0           0  \n",
       "1          47         28         17          1          0           0  \n",
       "2           8         21         26          1          0           0  \n",
       "3           7         15         17          5          1           0  \n",
       "4           0         12         14          9          1           0  \n",
       "5           1          4          4         19          1           1  \n",
       "6           0          0          5         27          2           0  \n",
       "7           0          0          1         17         11          11  \n",
       "8           0          0          0          9         16          24  \n",
       "9           0          0          0          6         13          23  \n",
       "10          0          0          0          1         14           5  \n",
       "11          0          0          0          0         10           8  \n",
       "12          0          0          0          0         11           6  \n",
       "13          0          0          0          0          2           3  \n",
       "14          0          0          0          0          1           1  \n",
       "15          0          0          0          0          1           1  \n",
       "16          0          0          0          0          0           1  \n",
       "17          0          0          0          0          1           0  \n",
       "18          0          0          0          0          0           0  \n",
       "19          0          0          0          0          0           0  \n",
       "20          0          0          0          0          0           0  \n",
       "21          0          0          0          0          0           0  \n",
       "22          0          0          0          0          0           0  \n",
       "23          0          0          0          0          0           0  \n",
       "24          0          0          0          0          0           0  \n",
       "25          0          0          0          0          0           0  \n",
       "26          0          0          0          0          0           0  \n",
       "27          0          0          0          0          0           0  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = OrderedDict()\n",
    "\n",
    "for fname in column_order:\n",
    "    df[fname] = [dna_variant_counts[fname][i] for i in range(max_mutations +1)]\n",
    "    \n",
    "pd.DataFrame.from_dict(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein sequence variation\n",
    "\n",
    "-\tForward files contain AA 2-121\n",
    "-\tReverse files contain AA 122-236, also in forward read orientation; thus naming the files as reverse is misleading.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(('91', 's', 'C'), ('118', 's', 'G'), ('293', 's', 'C'), ('322', 's', 'C'), ('359', 's', 'G'))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(((41, 's', 'V'), (99, 's', 'T'), (109, 's', 'L'), (121, 's', 'R')),\n",
       " 'I41V/K99T/I109L/M121R')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test protein_diff function for numbering\n",
    "\n",
    "test_ref =  \"AAGAGCTTGTTTAAGGGGCCGCGTGATTACAACCCGATATCGAGCACCATTTGTCATTTGACGAATGAATCTGATGGGCACACAACATCGTTGTATGGTATTGGATTTGGTCCCTTCATCATTACAAACAAGCACTTGTTTAGAAGAAATAATGGAACACTGTTGGTCCAATCACTACATGGTGTATTCAAGGTCAAGAACACCACGACTTTGCAACAACACCTCATTGATGGGAGGGACATGATAATTATTCGCATGCCTAAGGATTTCCCACCATTTCCTCAAAAGCTGAAATTTAGAGAGCCACAAAGGGAAGAGCGCATATGTCTTGTGACAACCAACTTCCAAACTAAGAGCATG\"\n",
    "test_read = \"AAGAGCTTGTTTAAGGGGCCGCGTGATTACAACCCGATATCGAGCACCATTTGTCATTTGACGAATGAATCTGATGGGCACACAACATCGCTGTATGGTATTGGATTTGGTCCCTTCGTCATTACAAACAAGCACTTGTTTAGAAGAAATAATGGAACACTGTTGGTCCAATCACTACATGGTGTATTCAAGGTCAAGAACACCACGACTTTGCAACAACACCTCATTGATGGGAGGGACATGATAATTATTCGCATGCCTAAGGATTTCCCACCATTTCCTCAAAAGCTGACATTTAGAGAGCCACAAAGGGAAGAGCGCCTATGTCTTGTGACAACCAACTTCCAAACTAAGAGCAGG\"\n",
    "\n",
    "print(find_dna_subs(test_read, test_ref))\n",
    "find_protein_diff(test_read, test_ref, start_offset = -3, end_trail = 0)\n",
    "\n",
    "# since first AA should be position 2, use start_offset -3\n",
    "# I need to disable end matching from the find_protein_diff function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_protein_diff(read, ref, verbose=False, start_offset=3, end_trail=3):\n",
    "\n",
    "    # quality control\n",
    "    if read is None:\n",
    "        return None, None\n",
    "    ends = findEnds(read, ref, start_offset)\n",
    "#     if not endMatch(read, ref, ends):\n",
    "#         return None, None\n",
    "\n",
    "    newread = read\n",
    "    newref = ref\n",
    "\n",
    "    # scan reference triplet by triplet\n",
    "    # move letters when encountering an indel\n",
    "    prot_errors = []\n",
    "    prot_short = []\n",
    "\n",
    "    i = ends.get('aligned')\n",
    "    ref_index = int((ends.get('aligned') - start_offset)/3) + 1  # reference amino acid index\n",
    "    max_i = len(ref) - end_trail\n",
    "\n",
    "    while i <= ends.get('end'):\n",
    "        if i > max_i:\n",
    "            break\n",
    "\n",
    "        if newread is None:\n",
    "            break\n",
    "        ref_codon = newref[i:i+3]\n",
    "        read_codon = newread[i:i+3]\n",
    "\n",
    "        if '-' in read_codon:  # found a deletion\n",
    "            # Check if this is the last acid, and it's incomplete, ignore it.\n",
    "            if re.search('[ATGC]', str(newread[i + 3:])) is None:\n",
    "                break\n",
    "\n",
    "            if '-' in ref_codon:  # something very broken\n",
    "                prot_errors.append((ref_index, 'f'))\n",
    "                prot_short.append('f')\n",
    "                break\n",
    "            elif read_codon == '---':  # single codon deletion\n",
    "                if ref_index > 0:\n",
    "                    prot_errors += [(ref_index, 'd')]\n",
    "                    prot_short.append(str(ref_index) + 'Î”')\n",
    "                i += 3\n",
    "                ref_index += 1\n",
    "\n",
    "            else:  # check it's not a frame shift\n",
    "                l = indel_len(newread, i)\n",
    "                if l % 3 != 0:\n",
    "                    prot_errors.append((ref_index, 'f'))\n",
    "                    prot_short.append('f')\n",
    "                    break\n",
    "                # realign gap and repeat loop at same position to compare the codons\n",
    "                gap = findGap(newread[i - 1:])\n",
    "                gap = (gap[0] + i - 1, gap[1] + i - 1)\n",
    "                newread = gapAlign(newread, gap, start_offset)\n",
    "                continue\n",
    "\n",
    "        elif '-' in ref_codon:  # found an insertion\n",
    "            l = indel_len(newref, i)\n",
    "            if l % 3 != 0:\n",
    "                prot_errors.append((ref_index, 'f'))\n",
    "                prot_short.append('f')\n",
    "                break\n",
    "            gap = findGap(newref[i-1:])\n",
    "            if gap[0] == 1:  # insertion after codon\n",
    "                insertion = newread[gap[0] + i - 1:gap[1] + i - 1]\n",
    "                if '-' in insertion:\n",
    "                    prot_errors.append((ref_index, 'f'))\n",
    "                    prot_short.append('f')\n",
    "                    break\n",
    "                if ref_index > 0:\n",
    "                    prot_errors.append((ref_index - 1, 'i', str(translate(insertion)))) # position before + insertion\n",
    "                    stop, inslist = format_insertion(ref_index - 1, insertion)\n",
    "                    prot_short += inslist\n",
    "                    if stop:\n",
    "                        break\n",
    "                i += l\n",
    "                ref_index += 1\n",
    "            else:  # realign gap and repeat loop at same position to compare the codons\n",
    "                gap = (gap[0] + i - 1, gap[1] + i - 1)\n",
    "                newref = gapAlign(newref, gap, start_offset)\n",
    "                continue\n",
    "\n",
    "        elif translate(read_codon) != translate(ref_codon):  # must be a substitution\n",
    "            if ref_index > 0:\n",
    "                prot_errors.append((ref_index, 's', str(translate(read_codon))))\n",
    "                prot_short.append(str(translate(ref_codon) + str(ref_index) + str(translate(read_codon))))\n",
    "            if str(translate(read_codon)) == '*':\n",
    "                break\n",
    "            i += 3\n",
    "            ref_index += 1\n",
    "\n",
    "        else:\n",
    "            i += 3\n",
    "            ref_index += 1\n",
    "\n",
    "    if verbose:\n",
    "        print(prot_errors)\n",
    "\n",
    "    if prot_short == []:\n",
    "        short = 'wt'\n",
    "    else:\n",
    "        short = '/'.join(prot_short)\n",
    "\n",
    "    return tuple(prot_errors), short\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's analyse to all second half sequences\n",
    "with open(\"./TEV_rv.fa\", \"r\") as f:\n",
    "    ref_seq =  SeqIO.read(f, \"fasta\", alphabet=IUPAC.ambiguous_dna)\n",
    "\n",
    "protein_variant_counts = {}\n",
    "\n",
    "files = glob.glob(\"/home/mp/InDelScanner/input_fasta/*_rv*\")\n",
    "\n",
    "for f in files:\n",
    "    # make a dictionary to collect the counts\n",
    "    fname = f[34:]\n",
    "    protein_variant_counts[fname] = defaultdict(int)\n",
    "    \n",
    "    for var_record in SeqIO.parse(f, \"fasta\", alphabet=IUPAC.ambiguous_dna):\n",
    "        dna_d = find_dna_subs(var_record.seq, ref_seq.seq, start_offset = 0, end_trail = 0)\n",
    "        if dna_d is None:\n",
    "            print(var.record)\n",
    "        else:\n",
    "            protein_d, prot_short = find_protein_diff(var_record.seq, ref_seq.seq, start_offset = -3, end_trail = 0)\n",
    "            protein_variant_counts[fname][len(protein_d)] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dna_variant_counts' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-66366c146f53>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdna_variant_counts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'dna_variant_counts' is not defined"
     ]
    }
   ],
   "source": [
    "dna_variant_counts.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['H_rv_1.fa', 'H_rv_11.fa', 'H_rv_3.fa', 'H_rv_5.fa', 'H_rv_7.fa', 'H_rv_9.fa', 'L_rv_1.fa', 'L_rv_11.fa', 'L_rv_3.fa', 'L_rv_5.fa', 'L_rv_7.fa', 'L_rv_9.fa', 'TEV_rv.fa'])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_variant_counts.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_mutations = max(list(itertools.chain.from_iterable(protein_variant_counts))) #  this tolerates an empty input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '1',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '1',\n",
       " '1',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '3',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '5',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '7',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'H',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '9',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '1',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '1',\n",
       " '1',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '3',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '5',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '7',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'L',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '_',\n",
       " '9',\n",
       " '.',\n",
       " 'f',\n",
       " 'a',\n",
       " 'T',\n",
       " 'E',\n",
       " 'V',\n",
       " '_',\n",
       " 'r',\n",
       " 'v',\n",
       " '.',\n",
       " 'f',\n",
       " 'a']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(itertools.chain.from_iterable(protein_variant_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_errors = find_dna_diff(read, ref, debug, start_offset, end_trail)  # errors = a tuple\n",
    "dna_hgvs = find_dna_hgvs(read, ref, refname, debug, start_offset, end_trail)  # string in HGVS format (ish)\n",
    "prot_errors, prot_short = find_protein_diff(read, ref, debug, start_offset, end_trail)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
